presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.orc.reader;
+
+import com.facebook.presto.orc.OrcCorruptionException;
+import com.facebook.presto.orc.StreamDescriptor;
+import com.facebook.presto.orc.stream.InputStreamSource;
+import com.facebook.presto.orc.stream.InputStreamSources;
+import com.facebook.presto.orc.stream.LongInputStream;
+
+import java.io.IOException;
+
+import static com.facebook.presto.orc.metadata.Stream.StreamKind.DICTIONARY_DATA;
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class LongDictionaryProvider
+{
+    public LongDictionaryProvider(InputStreamSources dictionaryStreamSources)
+    {
+        this.dictionaryStreamSources = dictionaryStreamSources;
+    }
+
+    public long[] getDictionary(StreamDescriptor streamDescriptor, int items)
+            throws IOException
+    {
+        // Return as is if not a flatmap subcolumn or that dictionary is not shared.
+        InputStreamSource<LongInputStream> dictionaryDataStream = dictionaryStreamSources.getInputStreamSource(streamDescriptor, DICTIONARY_DATA, LongInputStream.class);
+        checkArgument(streamDescriptor.getSequence() == 0 || dictionaryDataStream != null, "Missing dictionary stream");
+        return loadDictionary(streamDescriptor, dictionaryDataStream, items);
+    }
+
+    // We construct and use the input stream exactly once per stream description per stripe, so we don't
+    // really need to cache it.
+    private long[] loadDictionary(StreamDescriptor streamDescriptor, InputStreamSource<LongInputStream> dictionaryDataStream, int items)
+            throws IOException
+    {
+        try {
+            long[] dictionaryBuffer = new long[items];
+            LongInputStream inputStream = dictionaryDataStream.openStream();
+            if (inputStream == null) {
+                throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Dictionary is not empty but data stream is not present");
+            }
+            inputStream.next(dictionaryBuffer, items);
+            return dictionaryBuffer;
+        }
+        catch (java.lang.ClassCastException e) {
```

Comment:
What is the reason for ClassCastException and what source throws this error.

Reply:
This is thrown when trying to cast check stream point to a different type. I have a test case covering specifically this case to catch trying to get string dictionary buffer from LongDictionaryProvider.

Reply:
Can you provide the call stack when such an exception happens ? It is not clear to me why we need to handle this exception. I checked for String match of ClassCastException ini presto-orc and there is none.

Reply:
This doesn't currently happen. I'm adding this check just so that during such cases we have sane exception handling. If it does happen, it is programming error.